### PR TITLE
JLI Support

### PR DESCRIPTION
**Pull request by jm4rtin**
This change adds JLI support to GCC. This will accept the jli_call_fixed and jli_call_always pragmas used by Metaware. This implementation should be binary compatible with Metaware (if using the default option -mjli-func-addr=compat).

There are three approaches to obtaining the address of a function inside the JLI table:
1. -mjli-func-addr=compat will use a Metaware compatible mode. When using jli_call_fixed this will use the address of the function originally placed into the table at compile time. When using jli_call_always this will use the address of the JLI entry. The JLI table may not be moved when using jli_call_always with this method.
2. -mjli-func-addr=init will read the jli_base AUX register and offset it to determine the address of the JLI entry. This address will be stored in the desired variable. If the variable is constant or otherwise ends up in the .rodata section, the address of the .jlitab section will be used in place of the jli_base AUX register. This will emit an ARC_32_JLI relocation entry that may be used to perform the proper relocation utilizing the jli_base AUX register during loading. This method requires the call be made in kernel mode (because of the AUX register read).
3. -mjli-func-addr=always will create a stub function that reads the jli_base AUX register and offset to determine the address of the JLI entry. The address of the stub function will be stored in the desired variable so the JLI table may be moved after the variable has been set with the function address. This method requires the call be made in kernel mode (because of the AUX register read).

The following relocation types were added to GCC/binutils:
1. ARC_JLI_SECTOFF is a relocation type in Metaware that is now used by GCC as well to adjust the index of function calls to functions defined with #pragma jli_call_always during linking.
2. ARC_JLI_SECTOFF_SIMM12 is used to add the appropriate offset to the address when using -mjli-func-addr=init to get an address for a function in the JLI table during code execution.
3. ARC_JLI_32 is used to add the appropriate offset to the address when using -mjli-func-addr=init to get an address for a function in the JLI table during program load.

This implementation of the JLI feature will allow the use of both jli_call_fixed and jli_call_always as long as all jli_call_fixed functions are using the continuous block starting at index 0 and are linked before all other JLI entries. This can be done with the default linker script if the object file that contains the jli_call_fixed entries matches the object filename pattern "jlitab*.o".

### CODE REVIEW

**Comment by claziss** [comment]
Thank you for ur contribution. I will make a first review using your commits. Please let us know if you would like to take those in or you would like us to clean it up.

File: gcc/config/arc/arc-c.c [hunk 1]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
     builtin_define ("__big_endian__");
 
 }
+
+void
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:113]: Missing (a short) function description.

File: gcc/config/arc/arc-c.c [hunk 2]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+  enum cpp_ttype type;
+  char *symbol;
+  int index;
+
+  type = pragma_lex (&x);
+
+  if (type != CPP_OPEN_PAREN)
+  {
+    error ("expected '(' in #pragma jli_call_fixed");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:126]: extra space

File: gcc/config/arc/arc-c.c [hunk 3]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    return;
+  }
+
+  type = pragma_lex (&x);
+
+  if (type != CPP_NAME)
+  {
+    error ("expected function name in #pragma jli_call_fixed");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:135]: extra space

File: gcc/config/arc/arc-c.c [hunk 4]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+  }
+
+  symbol = strndup (IDENTIFIER_POINTER (x), 2048);
+
+  type = pragma_lex (&x);
+
+  if (type != CPP_COMMA)
+  {
+    free (symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:146]: extra space

File: gcc/config/arc/arc-c.c [hunk 5]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    error ("expected ',' in #pragma jli_call_fixed");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:148]: extra space

File: gcc/config/arc/arc-c.c [hunk 6]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    return;
+  }
+
+  type = pragma_lex (&x);
+
+  if (type != CPP_NUMBER || TREE_CODE (x) != INTEGER_CST)
+  {
+    free (symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:157]: extra space

File: gcc/config/arc/arc-c.c [hunk 7]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    error ("expected integer JLI index in #pragma jli_call_fixed");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:159]: extra space

File: gcc/config/arc/arc-c.c [hunk 8]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    return;
+  }
+
+  index = TREE_INT_CST_LOW (x);
+
+  if (0 > index || index >= ARC_JLI_ENTRIES_MAX)
+  {
+    free (symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:168]: extra space

File: gcc/config/arc/arc-c.c [hunk 9]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    error ("out of range JLI index in #pragma jli_call_fixed");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:170]: extra space

File: gcc/config/arc/arc-c.c [hunk 10]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    return;
+  }
+
+  type = pragma_lex (&x);
+
+  if (type != CPP_CLOSE_PAREN)
+  {
+    free (symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:179]: same here

File: gcc/config/arc/arc-c.c [hunk 11]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    error ("expected ')' in #pragma jli_call_fixed");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:181]: extra space

File: gcc/config/arc/arc-c.c [hunk 12]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+  {
+    free (jli_fixed_table[index]);
+
+    warning (OPT_Wpragmas, "function '%s' is already in the JLI table so the "
+      "index will be replaced with %d", symbol, index);
+  }
+  else if ((jli_fixed_count + jli_dynamic_count) >= ARC_JLI_ENTRIES_MAX)
+  {
+    free (symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:195]: extra space

File: gcc/config/arc/arc-c.c [hunk 13]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    error ("there are too many JLI entries");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:197]: extra space

File: gcc/config/arc/arc-c.c [hunk 14]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+  }
+  else
+  {
+    jli_dynamic_count++;
+  }
+
+  jli_fixed_table[index] = symbol;
+}
+
+void
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:208]: Missing (a short) function description

File: gcc/config/arc/arc-c.c [hunk 15]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+  tree x;
+  enum cpp_ttype type;
+  char *symbol;
+
+  type = pragma_lex (&x);
+
+  if (type != CPP_OPEN_PAREN)
+  {
+    error ("expected '(' in #pragma jli_call_always");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:220]: extra space

File: gcc/config/arc/arc-c.c [hunk 16]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    return;
+  }
+
+  type = pragma_lex (&x);
+
+  if (type != CPP_NAME)
+  {
+    error ("expected function name in #pragma jli_call_always");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:229]: extra space

File: gcc/config/arc/arc-c.c [hunk 17]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    return;
+  }
+
+  symbol = strndup (IDENTIFIER_POINTER (x), 2048);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:234]: extra space

File: gcc/config/arc/arc-c.c [hunk 18]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+  type = pragma_lex (&x);
+
+  if (type != CPP_CLOSE_PAREN)
+  {
+    free (symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:240]: extra space

File: gcc/config/arc/arc-c.c [hunk 19]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    error ("expected ')' in #pragma jli_call_always");
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:242]: extra space

File: gcc/config/arc/arc-c.c [hunk 20]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    return;
+  }
+
+  if (arc_jli_fixed_symbol_index (symbol) >= 0 ||
+    arc_jli_dynamic_symbol_index (symbol) >= 0)
+  {
+    free (symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:250]: extra space

File: gcc/config/arc/arc-c.c [hunk 21]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    warning (OPT_Wpragmas, "'%s' has already been registered as a JLI "
+      "function; ignoring", symbol);
+
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:253]: extra space

File: gcc/config/arc/arc-c.c [hunk 22]
```diff
@@ -106,3 +109,174 @@ arc_cpu_cpp_builtins (cpp_reader * pfile)
...
+    error ("there are too many JLI entries");
+
+    return;
+  }
+
+  jli_dynamic_table[jli_dynamic_count++] = symbol;
+}
+
+void
```
Comment (claziss) [inline gcc/config/arc/arc-c.c:269]: missing a short function description

File: gcc/config/arc/arc-opts.h [hunk 23]
```diff
@@ -91,4 +91,12 @@ enum arc_abi_type
 /* Default MPY option value.  */
 #define DEFAULT_arc_mpy_option -1
 
+/* Ways to determine the address of a function in the JLI table. */
```
Comment (claziss) [inline gcc/config/arc/arc-opts.h:94]: After dot you need to add two spaces

File: gcc/config/arc/arc.c [hunk 24]
```diff
@@ -69,6 +69,24 @@ along with GCC; see the file COPYING3.  If not see
 static char arc_cpu_name[10] = "";
 static const char *arc_cpu_string = arc_cpu_name;
 
+typedef struct _arc_jli_section
```
Comment (claziss) [inline gcc/config/arc/arc.c:72]: Please be a bit more verbose and describe those structures

File: gcc/config/arc/arc.c [hunk 25]
```diff
@@ -5068,6 +5091,57 @@ static void arc_file_start (void)
   fprintf (asm_out_file, "\t.cpu %s\n", arc_cpu_string);
 }
 
+static void arc_file_end (void)
```
Comment (claziss) [inline gcc/config/arc/arc.c:5094]: missing a short description

File: gcc/config/arc/arc.c [hunk 26]
```diff
@@ -5068,6 +5091,57 @@ static void arc_file_start (void)
...
+    fprintf (asm_out_file, "#####################\n");
+    fprintf (asm_out_file, "# JLI entry for function '%s'\n", sec->name);
+    fprintf (asm_out_file, "#####################\n");
+    fprintf (asm_out_file, "\t.section .jlitab$%s, \"ax\", @comdat\n",
+      sec->name);
+
+    if(first)
+    {
+      fprintf (asm_out_file, "\t.reloc 0, R_ARC_NONE, _init_jli\n");
+
```
Comment (claziss) [inline gcc/config/arc/arc.c:5112]: extra space

File: gcc/config/arc/arc.c [hunk 27]
```diff
@@ -7893,6 +7967,123 @@ compact_sda_memory_operand (rtx op, enum machine_mode mode)
...
+static bool
+arc_assemble_integer (rtx value, unsigned int size, int aligned_p)
+{
+  if (arc_is_call_to_jli_function (value))
+  {
+    const char *jli_symbol = XSTR (value, 0);
+
+    switch (arc_jli_func_addr)
+    {
+      // -mjli-func-addr=compat
```
Comment (claziss) [inline gcc/config/arc/arc.c:7979]: C++ comments, you need to use C comments

File: gcc/config/arc/arc.c [hunk 28]
```diff
@@ -7893,6 +7967,123 @@ compact_sda_memory_operand (rtx op, enum machine_mode mode)
...
+      case ARC_FUNC_ADDR_COMPAT:
+      {
+        // jli_call_always uses the address of the JLI entry and
```
Comment (claziss) [inline gcc/config/arc/arc.c:8036]: C++ comments

File: gcc/config/arc/arc.c [hunk 29]
```diff
@@ -12545,4 +12736,214 @@ insn_is_tls_gd_dispatch (rtx insn)
 
 struct gcc_target targetm = TARGET_INITIALIZER;
 
+/** Number of JLI entries in the fixed table. */
```
Comment (claziss) [inline gcc/config/arc/arc.c:12739]: One extra *, two spaces after dot.

File: gcc/config/arc/arc.h [hunk 30]
```diff
@@ -1687,4 +1675,26 @@ enum
 			  && (arc_fpu_build == FPX_QK))
 #define TARGET_FP_ASSIST ((arc_fpu_build & FPX_DP) != 0)
 
+/* Handle pragmas for JLI calls. */
+#define REGISTER_TARGET_PRAGMAS() do {                                  \
+  c_register_pragma (0, "jli_call_fixed",  arc_pr_jli_call_fixed);      \
```
Comment (claziss) [inline gcc/config/arc/arc.h:1680]: I am a bit reluctant in using pragmas with gcc compiler. I am wondering if it is really necessary. I know the MWDT use them quite extensively. But this is not the case of gcc.

File: gcc/config/arc/arc.md [hunk 31]
```diff
@@ -709,11 +712,126 @@
...
+    (unspec_volatile [
+      (match_operand:SI 0 "move_dest_operand" "")
+      (match_operand:SI 1 "symbolic_operand" "")
+    ] VUNSPEC_ARC_ADD2_JLIOFF)
+  ]
+  ""
+  "add2_jlioff %0, %0, %1"
+  [(set_attr "type" "unary")
+  (set_attr "iscompact" "true")
+  (set_attr "predicable" "yes")
```
Comment (claziss) [inline gcc/config/arc/arc.md:726]: Probably you want to say "predicable" "no" as the instruction string doesn't print any predication.

File: gcc/config/arc/arc.md [hunk 32]
```diff
@@ -709,11 +712,126 @@
...
+
+(define_insn_and_split "movsi_jli"
+  [(unspec_volatile [
+    (match_operand:SI 0 "move_dest_operand" "")
+    (match_operand:SI 1 "symbolic_operand" "")
+   ] VUNSPEC_ARC_MOVSI_JLI)
+   (clobber (match_scratch:SI 2 "=r"))
+  ]
+  ""
+  "lr %0, [JLI_BASE]
```
Comment (claziss) [inline gcc/config/arc/arc.md:737]: Probably you would like to use #.

File: gcc/config/arc/arc.md [hunk 33]
```diff
@@ -709,11 +712,126 @@
...
+    (set (match_dup 0) (match_dup 2))
+  ]
+  "")
+
 (define_expand "movsi"
   [(set (match_operand:SI 0 "move_dest_operand" "")
 	(match_operand:SI 1 "general_operand" ""))]
   ""
-  "if (prepare_move_operands (operands, SImode)) DONE;")
+  "// Use this to determine if the default action should be taken.
```
Comment (claziss) [inline gcc/config/arc/arc.md:752]: C++ comments. Please move all the logic into a new function in arc.c

File: gcc/config/arc/arc.md [hunk 34]
```diff
@@ -3978,6 +4096,45 @@
    (set_attr "iscompact" "true")
    (set_attr "cond" "nocond")])
 
+(define_insn "jli_s"
+  [(match_operand:SI 0 "immediate_operand" "L")]
+  ""
```
Comment (claziss) [inline gcc/config/arc/arc.md:4101]: I think jli is only valid for ARCv2 processors. I miss this condition in your code.

File: gcc/config/arc/arc.md [hunk 35]
```diff
@@ -3978,6 +4096,45 @@
...
+  "jli_s %0"
+  [(set_attr "type" "call")
+  (set_attr "iscompact" "true")
+  (set_attr "predicable" "yes")
```
Comment (claziss) [inline gcc/config/arc/arc.md:4105]: Probably you want to say "predicable" "no" as the instruction string doesn't print any predication.
 Add a line note

File: gcc/config/arc/arc.opt [hunk 36]
```diff
@@ -102,6 +102,25 @@ Same as -mcpu=archs
 ; END backward-compatibility aliases, translated by DRIVER_SELF_SPECS
 
 TargetVariable
+unsigned int arc_jli_func_addr = 0
+
+mjli-func-addr=
+Target RejectNegative Joined Var(arc_jli_func_addr) Enum(arc_jli_func_addr_type) Init(ARC_FUNC_ADDR_COMPAT)
+Specify the way the address of a function in the JLI table is obtained. Option 'compat' will use the Metaware method (jli_call_always will use the address of the function and jli_call_dynamic will use the address of the JLI entry). Option 'init' will use the address of the JLI entry using the jli_base register when the function pointer is saved. Option 'always' will use a stub function to call the JLI entry relative to the jli_base register upon calling the function pointed to.
```
Comment (claziss) [inline gcc/config/arc/arc.opt:109]: Here should be a short description. The long description should go into backend documentation file.

File: gcc/config/arc/arc.opt [hunk 37]
```diff
@@ -102,6 +102,25 @@ Same as -mcpu=archs
 ; END backward-compatibility aliases, translated by DRIVER_SELF_SPECS
 
 TargetVariable
+unsigned int arc_jli_func_addr = 0
```
Comment (claziss) [inline gcc/config/arc/arc.opt:105]: It is not necessary to declare arc_jli_func_addr

File: gcc/config/arc/predicates.md [hunk 38]
```diff
@@ -64,11 +72,20 @@
 		    && arc_legitimate_constant_p (VOIDmode, op)"))
 )
 
+(define_predicate "call_jli_address_operand"
+  (match_operand 0 "symbolic_jli_operand")
```
Comment (claziss) [inline gcc/config/arc/predicates.md:76]: Probably u can do something nicer here.

File: gcc/config/arc/arc-c.c (hunk 1 above)
Comment (jm4rtin) [inline gcc/config/arc/arc-c.c:113]: I can add one.

File: gcc/config/arc/arc-c.c (hunk 2 above)
Comment (jm4rtin) [inline gcc/config/arc/arc-c.c:126]: Please be more clear what you mean. For the most part it follows the horrible GCC style.

File: gcc/config/arc/arc.c (hunk 27 above)
Comment (jm4rtin) [inline gcc/config/arc/arc.c:7979]: I can fix that and add a few more.

File: gcc/config/arc/arc.h (hunk 30 above)
Comment (jm4rtin) [inline gcc/config/arc/arc.h:1680]: The intent is for the code to be compatible with Metaware. It's not strictly necessary but is preferred.

File: gcc/config/arc/arc.md (hunk 31 above)
Comment (jm4rtin) [inline gcc/config/arc/arc.md:726]: Thank you. I'll fix that. Wrapping my head around the GCC machine description has been a challenge.

File: gcc/config/arc/arc.md (hunk 32 above)
Comment (jm4rtin) [inline gcc/config/arc/arc.md:737]: Do you mean on this line?
`lr %0, [#JLI_BASE]`

Something else?

File: gcc/config/arc/arc.md (hunk 35 above)
Comment (jm4rtin) [inline gcc/config/arc/arc.md:4105]: As stated before, I'll fix the prediction. Add a line note?

File: gcc/config/arc/arc.md (hunk 34 above)
Comment (jm4rtin) [inline gcc/config/arc/arc.md:4101]: I'll look at how to limit this to  ARCv2 processors.

File: gcc/config/arc/arc.opt (hunk 37 above)
Comment (jm4rtin) [inline gcc/config/arc/arc.opt:105]: Then when is it necessary? I see other options declare variables, hence the confusion.

File: gcc/config/arc/predicates.md (hunk 38 above)
Comment (jm4rtin) [inline gcc/config/arc/predicates.md:76]: I'm open to suggestions on this matter.

**Comment by claziss** [comment]
PING: This patch needs to be modified to take advantage of the JLI support added into the binutils. Also we need to use function attributes instead of pragmas. Using pargmas makes this patch unlikely to be accepted into mainline GCC.

**Comment by fbedard** [comment]
Hi John,

Can you update us on your plans to update this and the corresponding binutils pull requests with review comments?  You mentioned your wish to see this included in our next release and we have about 2 weeks to include this if it can be modified per Claudiu's comments.   Thanks.

**Comment by jm4rtin** [comment]
Hi Francois,

This is an important feature we wish to add to the GCC port; however, we are unable to complete it within the two weeks allotted. As such, we would like to push this feature to the next release that I believe will be in December (2016.12). Thanks.

**Comment by claziss** [comment]
This entire patch is wrong. I'll try to distill it into a new implementation. I'll close pull request when the other one is done.

**Comment by claziss** [comment]
New jli support available in 2017.03. Closing

**Pull request closed without merging**